addons/binding/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/protocol/RemoteControllerWebSocket.java

```diff
@@ -0,0 +1,612 @@
+        json.put("id", uuid.toString());
+        json.put("params", jsonParams);
+
+        webSocketV2.sendCommand(json.toJSONString());
+    }
+
+    class App {
+        String appId;
+        String name;
+        int type;
+
+        App(String appId, String name, int type) {
+            this.appId = appId;
+            this.name = name;
+            this.type = type;
+        }
+
+        @Override
+        public String toString() {
+            return this.name;
+        }
+    }
+
+    Map<String, App> apps = new HashMap<>();
+
+    /**
+     * Send key code to Samsung TV.
+     *
+     * @param key Key code to send.
+     * @throws RemoteControllerException
+     */
+    @Override
+    public void sendKey(KeyCode key) throws RemoteControllerException {
+        sendKey(key, false);
+    }
+
+    public void sendKeyPress(KeyCode key) throws RemoteControllerException {
+        sendKey(key, true);
+    }
+
+    public void sendKey(KeyCode key, boolean press) throws RemoteControllerException {
+        logger.debug("Try to send command: {}", key);
+
+        if (!isConnected()) {
+            openConnection();
+        }
+
+        try {
+            sendKeyData(key, press);
+        } catch (RemoteControllerException e) {
+            logger.debug("Couldn't send command", e);
+            logger.debug("Retry one time...");
+
+            closeConnection();
+            openConnection();
+
+            sendKeyData(key, press);
+        }
+
+        logger.debug("Command successfully sent");
+    }
+
+    /**
+     * Send sequence of key codes to Samsung TV.
+     *
+     * @param keys List of key codes to send.
+     * @throws RemoteControllerException
+     */
+    @Override
+    public void sendKeys(List<KeyCode> keys) throws RemoteControllerException {
+        sendKeys(keys, 300);
+    }
+
+    /**
+     * Send sequence of key codes to Samsung TV.
+     *
+     * @param keys      List of key codes to send.
+     * @param sleepInMs Sleep between key code sending in milliseconds.
+     * @throws RemoteControllerException
+     */
+    public void sendKeys(List<KeyCode> keys, int sleepInMs) throws RemoteControllerException {
+        logger.debug("Try to send sequnce of commands: {}", keys);
+
+        if (!isConnected()) {
+            openConnection();
+        }
+
+        for (int i = 0; i < keys.size(); i++) {
+            KeyCode key = keys.get(i);
+            try {
+                sendKeyData(key, false);
+            } catch (RemoteControllerException e) {
+                logger.debug("Couldn't send command", e);
+                logger.debug("Retry one time...");
+
+                closeConnection();
+                openConnection();
+
+                sendKeyData(key, false);
+            }
+
+            if ((keys.size() - 1) != i) {
+                // Sleep a while between commands
+                try {
+                    Thread.sleep(sleepInMs);
+                } catch (InterruptedException e) {
```

Comment:
If we receive this InterruptedException, the interrupted flag was cleared and we most likely need to reset the flag:

http://www.yegor256.com/2015/10/20/interrupted-exception.html
https://www.ibm.com/developerworks/library/j-jtp05236/

Reply:
Thanks, will update.